src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,309 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+/**
+ * Builtin SASL OAUTHBEARER support
+ */
+#include "rdkafka_int.h"
+#include "rdkafka_transport_int.h"
+#include "rdkafka_sasl_int.h"
+
+
+/**
+ * @brief Per-connection state
+ */
+struct rd_kafka_sasl_oauthbearer_state {
+        enum {
+                RD_KAFKA_SASL_OAUTHBEARER_STATE_SEND_CLIENT_FIRST_MESSAGE,
+                RD_KAFKA_SASL_OAUTHBEARER_STATE_RECEIVE_SERVER_FIRST_MESSAGE,
+                RD_KAFKA_SASL_OAUTHBEARER_STATE_RECEIVE_SERVER_MESSAGE_AFTER_FAILURE,
+        } state;
+        rd_chariov_t server_error_msg;
+        /*
+         * A place to store a consistent view of the token and extensions
+         * throughout the authentication process -- even if it is refreshed
+         * midway through this particular authentication.
+         */
+        char *token_value;
+        char *md_principal_name;
+        rd_list_t extensions; /* rd_strtup_t list */
+};
+
+
+/**
+ * @brief Close and free authentication state
+ */
+static void rd_kafka_sasl_oauthbearer_close (rd_kafka_transport_t *rktrans) {
+        struct rd_kafka_sasl_oauthbearer_state *state = rktrans->rktrans_sasl.state;
+
+        if (!state)
+                return;
+
+        RD_IF_FREE(state->server_error_msg.ptr, rd_free);
+        rd_free(state->token_value);
+        rd_free(state->md_principal_name);
+        rd_list_destroy(&state->extensions);
+        rd_free(state);
+}
+
+
+
+/**
+ * @brief Build client-first-message
+ */
+static void
+rd_kafka_sasl_oauthbearer_build_client_first_message (
+        rd_kafka_transport_t *rktrans,
+        rd_chariov_t *out) {
+        struct rd_kafka_sasl_oauthbearer_state *state = rktrans->rktrans_sasl.state;
+
+        /*
+         * https://tools.ietf.org/html/rfc7628#section-3.1
+         * kvsep          = %x01
+         * key            = 1*(ALPHA)
+         * value          = *(VCHAR / SP / HTAB / CR / LF )
+         * kvpair         = key "=" value kvsep
+         * ;;gs2-header     = See RFC 5801
+         * client-resp    = (gs2-header kvsep *kvpair kvsep) / kvsep
+         */
+
+        const char *gs2_header = "n,,";
+        const char *kvsep = "\x01";
+        const int kvsep_size = strlen(kvsep);
+        int extension_size = 0;
+        int i;
+        for (i = 0 ; i < rd_list_cnt(&state->extensions) ; i++) {
+                rd_strtup_t *extension = rd_list_elem(&state->extensions, i);
+                // kvpair         = key "=" value kvsep
+                extension_size += strlen(extension->name) + 1 // "="
+                        + strlen(extension->value) + kvsep_size;
+        }
+
+        // client-resp    = (gs2-header kvsep *kvpair kvsep) / kvsep
+        out->size = strlen(gs2_header) + kvsep_size
+                + strlen("auth=Bearer ") + strlen(state->token_value) + kvsep_size
+                + extension_size + kvsep_size;
+        out->ptr = rd_malloc(out->size+1);
+
+        char *buf = out->ptr;
```

Comment:
put var defs at the beginning of the scope

Reply:
👍 Done (with all vars in this function)